functions.py

```diff
@@ -0,0 +1,192 @@
+'''
+
+import os
+import re
+import datetime
+import bugzilla
+from jira import JIRA
+
+
+def generate_html_file(htmlcode, remind=False):
+	''' generates HTML file of reminder
+	'''
+	try:
+		os.makedirs('archive')
+	except FileExistsError:
+		pass
+	if remind:
+		filename = './archive/reminder_{:%m%d%Y_%H:%M:%S}.html'.format(datetime.datetime.now())
+	else:
+		filename = './archive/report_{:%m%d%Y_%H:%M:%S}.html'.format(datetime.datetime.now())
+	with open(filename, 'w') as file:
+		file.write(htmlcode)
+
+
+def get_bugs_dict(bug_ids, config):
+	''' takes in set of bug_ids and returns dictionary with
+		bug_ids as keys and API data as values
+	'''
+
+	# initialize bug dictionary
+	bugs = {}
+
+	# iterate through bug ids from set
+	for bug_id in bug_ids:
+
+		# 0 should be default in YAML file (i.e. no bugs recorded)
+		# if present reference should be made in bug dict
+		if bug_id == 0:
+			bugs[0] = {'bug_name': 'No bug on file', 'bug_url': None}
+			continue
+
+		# get bug info from bugzilla API
+		try:
+
+			# hotfix: API call does not work if '/' present at end of URL string
+			parsed_bz_url = config['bz_url'].rstrip('/')
+
+			bz_api = bugzilla.Bugzilla(parsed_bz_url)
+			bug = bz_api.getbug(bug_id)
+			bug_name = bug.summary
+		except Exception as e:
+			print("Bugzilla API Call Error: ", e)
+			bug_name = "BZ#" + str(bug_id)
+		finally:
+			bug_url = config['bz_url'] + "/show_bug.cgi?id=" + str(bug_id)
+			bugs[bug_id] = {'bug_name': bug_name, 'bug_url': bug_url}
+
+	return bugs
+
+
+def get_bugs_set(blockers):
+	''' takes in blockers object and generates a set of all unique bug ids
+		including 0 if it is present
+	'''
+	bug_list = []
+	for job in blockers:
+		bz = blockers[job]['bz']
+		bug_list.extend(bz)
+
+	return set(bug_list)
+
+
+def get_jenkins_jobs(server, job_search_fields):
+	''' takes in a Jenkins server object and job_search_fields string
+		returns list of jobs with given search field as part of their name
+	'''
+
+	# parse list of search fields
+	fields = job_search_fields.split(',')
+
+	# fetch all jobs from server
+	all_jobs = server.get_jobs()
+
+	# parse out all jobs that do not contain any search field and/or are not OSP10, OSP13, OSP15 or OSP16 jobs
+	relevant_jobs = []
+	for job in all_jobs:
+		if ('10' not in job['name']) and ('13' not in job['name']) and ('15' not in job['name']) and ('16' not in job['name']):
+			continue
+		else:
+			for field in fields:
+				if field in job['name']:
+					relevant_jobs.append(job)
+					break
+
+	return relevant_jobs
+
+
+def get_jira_dict(ticket_ids, config):
+	''' takes in set of ticket_ids and returns dictionary with
+		ticket_ids as keys and API data as values
+	'''
+
+	# initialize ticket dictionary
+	tickets = {}
+
+	# initialize connection
+	auth = (config['jira_username'], config['jira_password'])
+	options = {
+		"server": config['jira_url'],
+		"verify": config['certificate']
+	}
+	jira = JIRA(auth=auth, options=options)
```

Comment:
Does the `JIRA` constructor need to be wrapped in a try/except block?

Reply:
Well, maybe that isn't necessary if the call site to this helper function does the error handling.

Reply:
It actually isn't - we have a try/except block for this in `get_bugs_dict()` but not here for some reason

@EliadCohen Did we have an architectural reason for this or is it just an oversight?

Reply:
Probably an oversight. It would be useful for several faults - credentials being wrong, temp outage on Jira. Will add issue for this

Reply:
Added #81